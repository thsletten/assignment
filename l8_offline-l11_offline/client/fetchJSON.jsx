export async function fetchJSON(url, options = {}){
    const res = await fetch(url, {
        method: options.method || "get",
        headers: options.json ? {"content-type" : "application/json"} : {},
        body: options.json && JSON.stringify(options.json),
    });

    if(!res.ok) {
        throw new Error(`Loading error: ${res.status} -> ${res.statusText};`)
    }

    if(res.status === 200){
        return await res.json();
    }
}