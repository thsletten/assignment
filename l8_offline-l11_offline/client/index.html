<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Back to the movies we go</title>
</head>
<body>
<div id="app">
</div>
</body>
<script src="index.jsx" type="module" />
</html>