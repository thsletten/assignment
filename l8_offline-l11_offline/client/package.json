{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "scripts": {
    "build": "npm install && parcel build index.html",
    "dev": "parcel watch index.html",
    "test": "jest",
    "verify": "jest --collectCoverage"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.19.4",
    "@babel/preset-react": "^7.18.6",
    "babel-jest": "^29.2.2",
    "jest": "^29.2.2",
    "jest-environment-jsdom": "^29.2.2",
    "jsdom": "^20.0.1",
    "parcel": "^2.7.0"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.2"
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ]
  },
  "jest": {
    "testEnvironment": "jsdom",
    "globals": {
      "IS_REACT_ACT_ENVIRONMENT": true
    },
    "collectCoverageFrom": [
      "**/*.{jsx,js}",
      "!index.jsx",
      "!coverage/**",
      "!dist/**"
    ]
  }
}
