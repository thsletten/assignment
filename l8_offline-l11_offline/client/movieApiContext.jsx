import React from "react";
import { fetchJSON } from "./fetchJSON.jsx";

export const MovieApiContext = React.createContext({
    async listMovies(title){
        return await fetchJSON(`/api/movies?titleSearch=${title}`);
    },
    async createMovie(movie){
      fetch("/api/movies", {
          method: "post",
          body: JSON.stringify(movie),
          headers: {
            "content-type": "application/json",
          },
      });
    },
});