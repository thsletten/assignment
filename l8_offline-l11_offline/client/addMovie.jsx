import * as React from "react";
import {useNavigate} from "react-router-dom";
import { useContext, useState } from "react";
import { MovieApiContext } from "./movieApiContext.jsx";


function FormInput({value, label, onChangeValue}) {
    return(
      <div>
          <label>
              <strong>{label} </strong>
              <input value={value} onChange={(e) => onChangeValue(e.target.value)} />
          </label>
      </div>
    );
}

export function AddMovie(){
    const { createMovie } = useContext(MovieApiContext);
    const navigate = useNavigate();

    const [title, setTitle] = useState("");
    const [yearInput, setYearInput] = useState("");
    const [plot, setPlot] = useState("");


    async function handleSubmit(e){
        e.preventDefault();
        const year = parseInt(yearInput);
        createMovie({ title, year, plot });
        navigate("/");
    }

    return(
        <form onSubmit={handleSubmit}>
            <h1> Submit new movie </h1>
            <FormInput label="Title:" value={title} onChangeValue={setTitle}/>
            <FormInput label="Year:" value={yearInput} onChangeValue={setYearInput}/>
            <FormInput label="Plot:" value={plot} onChangeValue={setPlot}/>
            <button> Submit </button>
        </form>
    );
}

