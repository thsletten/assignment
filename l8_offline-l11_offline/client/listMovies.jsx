import React, { useContext, useState } from "react";
import {MovieApiContext} from "./movieApiContext.jsx";
import {useLoader} from "./useLoader.jsx";

function MovieCard( { movie: { title, plot, year } }) {
    return(
        <>
            <h3> {title} - {year} </h3>
            <div> {plot} </div>
        </>
    );
}


export function ListMovies() {
    const { listMovies } = useContext(MovieApiContext);
    const [titleSearch, setTitleSearch] = useState("in");
    const [titleInput, setTitleInput] = useState("");
    const { loading, error, data } = useLoader(
        async () => await listMovies( titleSearch ),
        titleSearch
    );

    function handleSubmit(e){
        e.preventDefault();
        console.log("TitleInput: " + titleInput);
        setTitleSearch(titleInput);
    }

    if(loading) {
        return <div className="loading-indicator"> Still Loading... </div>;
    }

    if(error) {
        return (
            <div>
                <h1> Error </h1>
                <div className="error-message"> {error.toString()} </div>
            </div>
        );
    }

    return (
        <div>
            <h1> Movies to come back to: </h1>
            <div>
                <form onSubmit={handleSubmit}>
                    <label> Search by title:
                    <input
                        id="search-spot"
                        value={titleInput}
                        onChange={(e) => setTitleInput(e.target.value)}
                    />
                    <button> Search </button>
                    </label>
                </form>
            </div>
            {data.map((movie) => (
                <MovieCard key={movie.title} movie={movie} />
            ))}
        </div>
    );
}