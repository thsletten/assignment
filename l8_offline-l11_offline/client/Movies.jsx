import * as React from "react";
import {BrowserRouter, Link, Route, Routes } from "react-router-dom";

import { AddMovie } from "./addMovie.jsx";
import { ListMovies } from "./listMovies.jsx";

export function FrontPage() {
    return(
        <div>
            <h1> Back to the movies </h1>
            <ul>
                <li>
                    <Link to={"/movies/list"}> List Movies </Link>
                </li>
                <li>
                    <Link to={"/movies/new"}> Add new Movie </Link>
                </li>
            </ul>
        </div>
    );
}

export function Movies(){
    return(
        <Routes>
            <Route path={"/list"} element={<ListMovies />} />
            <Route path={"/new"} element={<AddMovie />} />
        </Routes>
    );
}

export function Application() {
    return(
        <BrowserRouter>
            <Routes>
                <Route path={"/"} element={<FrontPage />} />
                <Route path={"/movies/*"} element={<Movies />}/>
            </Routes>
        </BrowserRouter>
    );
}