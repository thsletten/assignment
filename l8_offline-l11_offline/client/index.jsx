import * as React from "react";
import { createRoot } from "react-dom/client";

import { Application } from "./Movies.jsx";

const element = document.getElementById("app");
const root = createRoot(element);

root.render(<Application />);