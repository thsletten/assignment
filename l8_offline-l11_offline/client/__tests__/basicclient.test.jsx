import * as React from "react"
import {createRoot} from "react-dom/client";
import { act } from "react-dom/test-utils";
import { MovieApiContext } from "../movieApiContext.jsx";

import { ListMovies } from "../Movies.jsx";

const movies = [
    {
        title: "Test movie 1",
        plot: "We test stuff",
        year: "2022"
    },
    {
        title: "Test movie 2",
        plot: "we test more stuff",
        year: "2022"
    }
]

async function renderListMovies(listMovies) {
    const element = document.createElement("div");
    const root = createRoot(element);

    await act(async () =>
        root.render(
            <MovieApiContext.Provider value={{ listMovies }}>
                <ListMovies />
            </MovieApiContext.Provider>
        )
    );
    return element;
}

describe("client test suite", () => {

    it("show loading screen", async () => {
        const element = await renderListMovies(() => new Promise(() => {}));

        expect(element.querySelector(".loading-indicator")).not.toBeNull();
        expect(element.innerHTML).toMatchSnapshot();
    });

    it("shows movies list", async () => {
        const element = await renderListMovies(async () => movies);

        expect(element.querySelector("h3").innerHTML).toEqual(` ${movies[0].title} - ${movies[0].year} `);
        expect(element.innerHTML).toMatchSnapshot();
    });

    it("shows error message", async () => {
        const element = await renderListMovies(async () => {
           throw new Error("Failed to fetch");
        });

        expect(element.querySelector(".error-message").innerHTML)
            .toContain("Failed to fetch");
        expect(element.innerHTML).toMatchSnapshot();

    });
});