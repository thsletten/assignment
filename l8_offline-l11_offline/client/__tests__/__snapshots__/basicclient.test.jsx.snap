// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client test suite show loading screen 1`] = `"<div class="loading-indicator"> Still Loading... </div>"`;

exports[`client test suite shows error message 1`] = `"<div><h1> Error </h1><div class="error-message"> Error: Failed to fetch </div></div>"`;

exports[`client test suite shows movies list 1`] = `"<div><h1> Movies to come back to: am I even looking right?</h1><h3> Test movie 1 - 2022 </h3><div> We test stuff </div><h3> Test movie 2 - 2022 </h3><div> we test more stuff </div></div>"`;
